### DISCUSSION TITLE
Roadmaps for improving quality?

### DISCUSSION THREAD

**Discussion by duanqn** [zxing-cpp/zxing-cpp#421] [Q&A]
Hi,
I found this to be a great barcode recognition library. I'd like to help improve it but I didn't see a contribution doc. Could the authors list some tasks / roadmaps for improving the recognition quality?

**Comment by huycn** [comment]
Thank you for your interests in contributing to this project. You can find the things need to be done in issue list. @axxel is leading, he will be more aware what things can be done or not. Otherwise, I would suggest go where the need is. If you find something this library lacks or it doesn't do well the job in some use case, make your change and submit your PR.

**Comment by axxel** [comment]
Also hi from my side and welcome. Out of curiosity I have two questions:
 a) What is your background and area of expertise?
 b) What is your motivation to contribute (do you have a specific use case of the library yourself or just want to contribute to an interesting open source project)?

I have several 'loose' ends in my head as to where the project can be improved but most of them are still relatively vague or rather long term changes. Depending on your answers I'll try to come up with a suggestion to start somewhere.

**Comment by duanqn** [comment]
a) I know C++ (I would say I'm a fair C++ programmer), but I don't have much knowledge in the specific domain of barcodes.
b) My immediate need is that I want to become more familiar with the recognition algorithms. Specifically, if someone asks me to explain why ZXing-cpp can / cannot recognize a barcode, I want to answer them without re-posting the question on GitHub. (Not all images can be posted on GitHub; Also I shouldn't expect open source maintainers to answer my questions 24x7)
Other than that, I'm happy to contribute to any tasks that may help this repo. It's always nice to get some experience in open source projects :)

**Comment by axxel** [comment]
You have a rather peculiar objective ;). So you are going to be the internal support for a presumably commercial environment?

If I'm confronted with this question (why is a particular image not recognized) I have to look at debugging output while running the code on the sample myself and I (re-)wrote half the detection algorithms in this code base. So I guess the best approach would likely be to just dig in once you have a specific sample to work with? As a little "exercise" you could start looking into the first 3 samples of #275 and see what your analysis is regarding the limitation of the current code.

If you want to tackle something 'big' where the current code is completely lacking then you might want to have a look at the Aztec and the MaxiCode detectors. Especially the latter is simply not working at all for real-world input. I have not looked at that code at all. The Aztec is also severely limited. There I do at least have some thoughts on how to possibly improve the situation. If you want to dive into any of those, I'd encourage you to make your progress visible, so you don't spent days or even weeks on that only to find out that I find your end result questionable for whatever reason. ;).

**Comment by duanqn** [comment]
Haha it's still in the early stage, not going commercial yet. I need to answer the questions from my supervisor.
Thanks for the advice. I'll dig deeper into the barcode readers.
